### PR TITLE
refactor: transition from open_ethereum primatives (types/rlp) to Reth's Alloy (types/rlp)

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?
- open ethereum has been deprecated for 4 years
- reth told us we should do this at Devconnect
- sigp in the process of updating from open_ethereum primatives (types/rlp) to Reth's Alloy (types/rlp). Currently sigp/(enr/discv) are updated, but lighthouse is planning to upgrade as well.
- Our current rlp code for era1 type transactions is broken, this PR will allow us to easily fix the the issue we have encoding typed transactions for era1 files
### How was it fixed?

update all our code to use Reth alloy types and rlp instead of open ethereum's

## why do we want this?
Open ethereum has been unmaintained for 4 years now. Switching to alloy (lighthouse is also planning to switch), will allow us to much more easily use libraries from Reth

A big reason is we can has a ton of crates we are going to want to use, this transition makes using Reth crates a billion percent easier.

### CODE REVIEW

**Comment by morph-dev** [comment]
There are several crates that are referencing your private repo. I checked few and they don't have only trivial changes.

Is there a plan for those crates to migrate to `alloy` as well, or do we plan to use alternative crates in the future? Would be good to know what are the next steps.

Maybe update the PR description with each such crate and plan for it for future.

**Comment by KolbyML** [comment]
> In general, I like the idea and the initiative. However, I have some concerns.
> 
> There are several crates that are referencing your private repo, meaning they wouldn't get any updates / bug fixes. I checked few and they don't have only trivial changes.

The only one that doesn't have trivial changes is eth-trie.rs which we maintain. That is because it also required me to switch updated its RLP. The rest of the libraries mentioned are maintained by sigma prime and the changes are trival just changing openethereum_types to alloy-primative types

The libraries maintained by sigma prime are
- ethereum_ssz
- ethereum_serde_utils
- tree_hash
- ssz_types

I made PR's for all 4, it took less then an hour combined to convert all of them from ethereum-types to Reth's alloy types. 

So yeah the only "private repo" with "don't have only trivial changes" is eth-trie.rs which we maintain. So I am not sure what "several" is referring to.
> Is there a plan for those crates to migrate to `alloy` as well, or do we plan to use alternative crates in the future? Would be good to know what are the next steps.

To my knowledge sigp is going to upgrade sigp/(ethereum_ssz/ethereum_serde_utils/tree_hash/ssz_types) I made PR's for it. They already upgraded sig/(enr/discv5). sigp/(ethereum_ssz/ethereum_serde_utils/tree_hash/ssz_types) are very stable and aren't in active development either, so I think we are good in that regard.

**Review by ogenev — CHANGES_REQUESTED** [review]
I have a few questions:

1. What is the motivation behind migrating from `ethereum_types` to `alloy-primitives`?

> sigp in the process of updating from open_ethereum primatives (types/rlp) to Reth's Alloy (types/rlp). Currently sigp/(enr/discv) are updated, but lighthouse is planning to upgrade as well.

2. Could you link to PR/issue about that. What I found is [this](https://github.com/sigp/discv5/pull/212) but they migrate only `rlp` to `alloy-rlp`.

3. Is it possible to migrate only the rlp crate and at a later stage to decide what to do with the primitive types?

**Comment by KolbyML** [comment]
> I have a few questions:
> 
> 1. What is the motivation behind migrating from `ethereum_types` to `alloy-primitives`?

image

Also so we can use the reth's REVM crate and json-rpc types seemlessly instead of a bunch of type conversion boilerplate.

> > sigp in the process of updating from open_ethereum primatives (types/rlp) to Reth's Alloy (types/rlp). Currently sigp/(enr/discv) are updated, but lighthouse is planning to upgrade as well.
> 
> 2. Could you link to PR/issue about that. What I found is [this](https://github.com/sigp/discv5/pull/212) but they migrate only `rlp` to `alloy-rlp`.

Discv5 doesn't use ethereum-types. I want to emphasize even if lighthouse aligns with this change or not that shouldn't impact the decision. We are an execution layer focused project and being able to use a lot of Reth's crates going forward will be a massive help. The 4 sigp libraries are very small, drops in a bucket compared to the REVM, even the json-rpc-types crate we want to use.

I asked and here is the response I got
image
image
So it looks like in a week or so after a team meeting we should get a definite answer. 
> 3. Is it possible to migrate only the rlp crate and at a later stage to decide what to do with the primitive types?

- Sigp's response referred to using "open-ethereum-types" as a bandaid in the image above
- In the Image from Dragon from Reth he referred to it as a liability.

The truth being Parity hasn't in the Ethereum space for 4 years now. So I think it is clear we want to use Reth's primatives expessially with our growing interest in the REVM I see this as invaluable.

To answer the question directly
image
I don't see value wrapping and the extra work of supporting openethereums-types as worth it. We would have to wrap all the types we use.

**Comment by ogenev** [comment]
I agree that this is a good direction to go forward but I think the switch should be done more carefully. 

This is a core change to our codebase and we should be more cautious relying on custom forks for core libraries like `ethereum_ssz`, `tree_hash`, `ethereum_serde_utils`,  `ssz_types` and etc. 

What problem are we trying to solve here? We want to fix an era1 decoding bug.  Is this the simplest solution to this problem? Is this change at the top list of our priorities?

I would wait for sgp to update their crates and in the meantime, I'm more comfortable applying a quick fix for the era1 decoding issue.

**Comment by morph-dev** [comment]
I would say I'm with @ogenev on this one.

I noticed that [`alloy-primitives`](https://docs.rs/crate/alloy-primitives/latest/features) has features that might be useful for us and maybe eliminate need to have custom changes: `ssz` and maybe `serde`. Did you look into those?

Would it be possible to migrate in smaller chunks? Crate-by-crate would probably not work, but maybe structure-by-structure, i.e. each subnetwork separately.
It's definitely not ideal to depend on multiple crates that do the same thing, but I would say it's fine if it is just temporary thing (e.g. until Sigp migrates their crates).

Another related topic
Changes in `eth_trie` are not trivial, and crate itself wasn't really designed for our use case.
If we decide not to proceed with this PR, I can put an effort into creating and publishing the crate that will be more suited for us, that will use `alloy` types and we can migrate that part separately.

**Comment by KolbyML** [comment]
> This is a core change to our codebase and we should be more cautious relying on custom forks for core libraries like `ethereum_ssz`, `tree_hash`, `ethereum_serde_utils`, `ssz_types` and etc.

Those libraries don't really change, they maybe add 1 thing every year or 6 months, and normally that is a new "feature" we don't need. The only change I made in those 4 libraries is cntr-alt-r open-ethereum types with reth's alloy types. This isn't the first time we have relied on custom forks well we are waiting for changes to be implemented on the main repo's. PR's are made on all the repos with the changes to reth's alloy types. On Thursday call if we want we can go over sigp's crates commit history to look at the changes. No meaningful changes have been made to those repos in years. ssz is simple serialization there isn't much to work on or change. I will take on the responsibility to work with sigp and maintain our forks for the forseeable future if required, till alloy is merged into them.

> What problem are we trying to solve here? We want to fix an era1 decoding bug. Is this the simplest solution to this problem with the least friction? Is the alloy types migration at the top list of our priorities?

This is the simplest solution as the work is already done and tested
- we want to use the revm in state
- we want to use reth-rpc-types (with this change we won't need a ton of boiler plate as we get endpoints online)
- Changing to this we can remove ethers-rs as a dependency as alloy replaces it as well
- fixing the era1 rlp encoding/decoding issue

After we launch history I will be going to work on the state network assuming everything goes to plan. REVM uses reth's alloy types/rlp not open-ethereums. This PR is the setup to work with Reth's crates. A majority of their work is in modular crates and switching to their types allows us to easily taken advantage of them.

> I would wait for sgp to update their crates before moving with this change and in the meantime, I'm more comfortable applying a quick fix for the era1 decoding issue.

This is the quick fix as the work is already done and tested. As stated above the sigp crates we rely on hardly change and when they do it is often a new "feature" we don't need. I am going to work with Sigp to resolve this for us, but I don't think this should block this PR.


**TLDR**
The work is already done. We want to switch because we want to use Reth's
- REVM crate for the state network
- json-rpc-types crate (without needing to write tons of boilerplate)
- verkle is coming up, we will want to use the same crate as Reth (our state network/Reth's execution code) go hand and hand, we don't want to rewrite everything for open-ethereum's types

That is to name a few examples, but this PR opens doors beyond the examples I gave today, parity hasn't been in the ethereum ecosystem for 4 years, we are an execution aligned project, it makes sense to change to these types so we can reuse Reth's execution layer code.

I know this PR is "big", but it is unavoidable and this technical debt only grows overtime. The work is done.

**Comment by KolbyML** [comment]
> Another related topic Changes in `eth_trie` are not trivial, and crate itself wasn't really designed for our use case. If we decide not to proceed with this PR, I can put an effort into creating and publishing the crate that will be more suited for us, that will use `alloy` types and we can migrate that part separately.

Are we not able to use this https://github.com/paradigmxyz/reth/tree/main/crates/trie ?

Reth has crates for all things execution layer, which is exactly why I think we should make this change. So we can reuse work already done instead of reinventing the wheel.

**Comment by morph-dev** [comment]
> Are we not able to use this https://github.com/paradigmxyz/reth/tree/main/crates/trie ?

I wasn't aware of that crate, probably because it's not published. I will look into it today.

**Comment by KolbyML** [comment]
It looks like Alex Stokes is maintaining an ssz library using Reth's alloy-primitives https://github.com/ralexstokes/ssz-rs so maybe the move would be to switch to that

**Comment by morph-dev** [comment]
> > Are we not able to use this https://github.com/paradigmxyz/reth/tree/main/crates/trie ?
> 
> I wasn't aware of that crate, probably because it's not published. I will look into it today.

I looked into it, and we can likely use it. But, it's not really created for our use case and it comes with extra cost. Currently, and I'm not sure this will change any time soon, we only need to traverse certain path in a trie and verify that result is what it needs to be.

And also, I'm not sure if we want to use `reth` crates. I think [`revm`](https://crates.io/crates/revm) might be enough for our use case (executing tx / block). The `revm` doesn't actually care about trie structure at all (at least to my understanding, it only needs a structure/db that will provide blocks, account info, code, and contract storage).

They all use alloy types, so transition would be useful anyway. All I'm saying is that we might not need to be so strongly integrated with `reth`.

File: trin-history/src/validation.rs [hunk 1]
```diff
@@ -216,7 +217,7 @@ mod tests {
     #[tokio::test]
     async fn validate_epoch_acc() {
         let epoch_acc =
-            std::fs::read("./../trin-validation/src/assets/epoch_accs/0x5ec1…4218.bin").unwrap();
+            std::fs::read("./../trin-validation/src/assets/epoch_accs/0x5ec1ffb8c3b146f42606c74ced973dc16ec5a107c0345858c343fc94780b4218.bin").unwrap();
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply trin-history/src/validation.rs:220]: open-ethereums display gives a short hash, where reth's alloy-primitive display returns the full hash

File: ethportal-api/src/types/execution/block_body.rs [hunk 2]
```diff
@@ -318,25 +304,7 @@ impl ssz::Decode for BlockBodyMerge {
...
-
-impl Decodable for BlockBodyMerge {
-    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
-        let txs: Vec<Transaction> = rlp.list_at(0)?;
-        Ok(Self { txs })
-    }
-}
-
-#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, RlpEncodable, RlpDecodable)]
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/block_body.rs:307]: This is changed in the era1 rlp fix PR

**Comment by KolbyML** [comment]
> Another related topic Changes in `eth_trie` are not trivial,

The only "non-trivial" change is in 1 function `decode_node()` and that change is updating it from using open-ethereum/rlp to alloy/rlp.  I want to clairfy this change isn't complex it is a library migration.

We can review the changes in this 1 function. I already tested the changes with our Portal-Hive state tests and the update to alloy/rlp passes all successfully. 

>. If we decide not to proceed with this PR, I can put an effort into creating and publishing the crate that will be more suited for us, that will use `alloy` types and we can migrate that part separately.

- We are moving away from open-ethereum types and rlp regardless. So it isn't a question of not getting this PR in, it is how we are updating away from open-ethereum types/rlp. And I don't see a small function we can review and has been extensively tested as a blocker for this PR. 

 I think writing a whole new merkle trie library is far more "non-trivial" then updating the rlp library eth-trie.rs uses and just reviewing it.

> Another related topic Changes in `eth_trie` are not trivial,

I am very confused how 1 function changing from open-ethereum/rlp to alloy/rlp is a more complex change then writing a brand new library.

Verkle is coming in a year? no? Won't that make a merkle based state network obsolete? So why does a library not designed for our specific temporary usecase which works for our usecase and works, justify building a special solution.

My main point is I don't think this blocks this PR. And I am not sure why the complexity of the changes in `eth-trie.rs` are being overblown to justify building a brand new `merkle trie` library. 

The changes in `eth-trie.rs` are not complex it is a library migration and the notable difference is 1 function which can be reviewed. It has already been tested and works with all our test data as well.

**Comment by KolbyML** [comment]
> And also, I'm not sure if we want to use `reth` crates. I think [`revm`](https://crates.io/crates/revm) might be enough for our use case (executing tx / block). The `revm` doesn't actually care about trie structure at all (at least to my understanding, it only needs a structure/db that will provide blocks, account info, code, and contract storage).

We already use
- reth's json-rpc-types (We already use this for get_block_by_hash, but with a lot of type conversion helper functions, we plan to support all ethereum json-rpc calls so reusing this will be super helper, changing to alloy means removing the boiler plate to use these types)
- We use there rpc server crate (this doesn't rely on alloy, but it is a reth crate)
- we will want to use the same verkle crate when that is made no?
- revm

If we don't want to use the `trie` crate that is fine, but deciding not to use 1 reth crate doesn't mean we don't get value from other `reth` crates or we won't use them.

> They all use alloy types, so transition would be useful anyway. All I'm saying is that we might not need to be so strongly integrated with `reth`.

Reth is modular. I am not sure how using a crate is the same thing as `All I'm saying is that we might not need to be so strongly integrated with reth` sentiment.

Crates are just tools, Reth just happens to build a lot of tools. If there is a tool we can use to get something done faster I see that as a win. We don't want to rebuild everything from scratch unless there is a reason for it.

File: portal-bridge/src/types/era1.rs [hunk 3]
```diff
@@ -196,7 +197,8 @@ impl TryInto<Entry> for HeaderEntry {
     type Error = anyhow::Error;
 
     fn try_into(self) -> Result<Entry, Self::Error> {
-        let rlp_encoded = rlp::encode(&self.header).to_vec();
+        let mut rlp_encoded = vec![];
+        self.header.encode(&mut rlp_encoded);
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply portal-bridge/src/types/era1.rs:201]: Very cool, I didn't see that I will try and apply this to the other places.

File: ethportal-api/src/types/consensus/serde.rs [hunk 4]
```diff
@@ -1,6 +1,7 @@
-use ethereum_types::U256;
+use alloy_primitives::U256;
 use serde::{ser::SerializeSeq, Deserialize, Deserializer, Serializer};
 use serde_json::Value;
+use serde_utils::u256_from_dec_str::u256_from_dec_str;
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/consensus/serde.rs:4]: It is in the PR u256_from_dec_str. https://github.com/sigp/ethereum_serde_utils/pull/7 I couldn't find it in alloy, so I just copy-pasted the utility function from open-ethereum to here as it was needed here.

I can make a PR for alloy-primitives to add this functionality natively, but this is a workaround for now.

File: ethportal-api/src/types/execution/block_body.rs [hunk 5]
```diff
@@ -165,22 +148,22 @@ impl BlockBody {
...
-        stream.append_list(&self.uncles()?);
-        let rlp = stream.out().freeze();
-
-        let hash = Keccak256::digest(&rlp);
-        Ok(H256::from_slice(&hash))
+    pub fn uncles_root(&self) -> anyhow::Result<B256> {
+        let mut buf = Vec::<u8>::new();
+        self.uncles()?.encode(&mut buf);
+        let hash = Keccak256::digest(&buf);
+        Ok(B256::from_slice(&hash[..32]))
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/block_body.rs:155]: Good catch we don't need to parse the value at all

File: ethportal-api/src/types/execution/block_body.rs [hunk 6]
```diff
@@ -190,28 +173,12 @@ impl BlockBody {
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, RlpEncodable, RlpDecodable)]
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/block_body.rs:176]: Same idea as https://github.com/ethereum/trin/pull/1231#discussion_r1550105854

**Comment by KolbyML** [comment]
I implemented the era1 change using open-ethereum/rlp
https://github.com/ethereum/trin/pull/1237
There is more information in the PR, but in short it is 3.64 times slower in an encode/decode test of an era1 file compared to using reth's alloy/rlp.

I don't think it makes sense to use open-ethereum rlp, for all the reasons above. 3.64 times slower is no joke.

File: ethportal-api/src/types/execution/transaction.rs [hunk 7]
```diff
@@ -346,31 +440,37 @@ impl<'de> Deserialize<'de> for ToAddress {
...
 
 impl Decodable for ToAddress {
-    fn decode(rlp: &Rlp<'_>) -> Result<Self, DecoderError> {
-        match rlp.is_empty() {
-            true => Ok(ToAddress::Empty),
-            false => Ok(ToAddress::Exists(rlp::decode(rlp.as_raw())?)),
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        if let Some(&first) = buf.first() {
+            if first == EMPTY_STRING_CODE {
+                buf.advance(1);
```
Comment (njgheorghita) [inline ethportal-api/src/types/execution/transaction.rs:467]: Why are we advancing the `buf` here? I'm just not sure that there's a test for this code path, or did you find one and this was required to get it passing?

File: ethportal-api/src/types/execution/block_body.rs [hunk 8]
```diff
@@ -355,14 +306,11 @@ impl ssz::Encode for BlockBodyShanghai {
...
-        let encoded_txs: Vec<Vec<u8>> = self.txs.iter().map(|tx| rlp::encode(tx).into()).collect();
+        let encoded_txs: Vec<Vec<u8>> = self.txs.iter().map(alloy_rlp::encode).collect();
         let empty_uncles: Vec<Header> = vec![];
-        let rlp_uncles: Vec<u8> = rlp::encode_list(&empty_uncles).to_vec();
-        let encoded_withdrawals: Vec<Vec<u8>> = self
-            .withdrawals
-            .iter()
-            .map(|withdrawal| rlp::encode(withdrawal).to_vec())
-            .collect();
+        let rlp_uncles: Vec<u8> = alloy_rlp::encode(empty_uncles);
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/block_body.rs:311]: Screenshot 2024-04-04 at 8 34 19 AM
we would then have to pass a mutiable vec, which isn't that unidle

File: ethportal-api/src/types/execution/header.rs [hunk 9]
```diff
@@ -90,128 +89,120 @@ where
...
-        } else {
-            13
-        };
-
-        if with_seal && self.mix_hash.is_some() && self.nonce.is_some() {
-            s.begin_list(stream_length_without_seal + 2);
-        } else {
-            s.begin_list(stream_length_without_seal);
+impl Encodable for Header {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/header.rs:98]: It wasn't used

File: ethportal-api/src/types/execution/receipts.rs [hunk 10]
```diff
@@ -418,42 +397,69 @@ impl Receipt {
             Self::Blob(receipt) => receipt,
         }
     }
+}
 
-    fn encode(&self) -> Vec<u8> {
-        let mut stream = RlpStream::new();
+impl Encodable for Receipt {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        let with_header = false;
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/receipts.rs:404]: It is what adds the rlp(txtype + rlp(tx)), with it false it is txtype + rlp(tx)

it is used in the follow up pr

File: ethportal-api/src/types/execution/transaction.rs [hunk 11]
```diff
@@ -18,50 +20,76 @@ pub enum Transaction {
...
+        self.encode(&mut buf);
+        keccak256(buf)
     }
 }
 
 impl Encodable for Transaction {
-    fn rlp_append(&self, s: &mut RlpStream) {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        // we don't wrap versioned transactions with a string header
+        let with_header = false;
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/src/types/execution/transaction.rs:33]: We use it in the era 1 fix PR https://github.com/ethereum/trin/pull/1231#discussion_r1551825455

File: ethportal-api/src/types/execution/transaction.rs (hunk 7 above)
Comment (KolbyML) [reply ethportal-api/src/types/execution/transaction.rs:465]: I think it is less error prone as well

**Review by njgheorghita — APPROVED** [review]

**Review by ogenev — APPROVED** [review]
Ok let's give this a go, main main comment is that I would like to see `alloy-primitives` version updated to the latest minor `0.7.0` on all crates. ⛵

File: ethportal-api/Cargo.toml [hunk 12]
```diff
@@ -29,23 +30,21 @@ lazy_static = "1.4.0"
 nanotemplate = "0.3.0"
 quickcheck = "1.0.3"
 rand = "0.8.5"
-reth-rpc-types = { tag = "v0.1.0-alpha.10", git = "https://github.com/paradigmxyz/reth.git"}
+reth-rpc-types = { tag = "v0.2.0-beta.4", git = "https://github.com/paradigmxyz/reth.git"}
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply ethportal-api/Cargo.toml]: Reth bumped the alloy version after the release of beta.5 so I updated it to the commit for the bump commit
